File: typescript-impl/src/infrastructure/repositories/bank_account_repository.ts
import { BankAccountEntity } from '../entities/bank_account_entity';
import { PersonEntity } from '../entities/person_entity';
import { AccountType } from '../enums/account_type';

export interface BankAccountRepository {
    find_by_id(bank_account_id: number): BankAccountEntity;
    create(account_holder: PersonEntity, initial_deposit: number, account_type: AccountType): BankAccountEntity;
    show_all_accounts(): BankAccountEntity[];
    update(bank_account_id: number, first_name: string, last_name: string, date_of_birth: Date, email: string, national_id: string, phone_number: string, parmanent_address: string): void;
    delete(bank_account_id: number): boolean;
    deposit(bank_account_id: number, amount: number): boolean;
    withdraw(bank_account_id: number, amount: number): boolean;
}

File: python-impl/insfrastructure/repositories/bank_account_repository.py
from abc import ABC, abstractmethod
from datetime import datetime

from insfrastructure.entities.bank_account_entity import BankAccountEntity
from insfrastructure.entities.person_entity import PersonEntity
from insfrastructure.enums.account_types import AccountType


class BankAccountRepository(ABC):
    @abstractmethod
    def find_by_id(self, bank_account_id: int) -> BankAccountEntity:
        pass
    
    @abstractmethod
    def create(self, account_holder: PersonEntity, initial_deposit: float, account_type: AccountType) -> BankAccountEntity:
        pass
    
    @abstractmethod
    def show_all_accounts(self) -> list[BankAccountEntity]:
        pass
    
    @abstractmethod
    def update(self, bank_account_id: int, first_name: str, last_name: str, date_of_birth: datetime, email: str, national_id: str, phone_number: str, parmanent_address: str):
        pass
    
    @abstractmethod
    def delete(self, bank_account_id: int) -> bool:
        pass
    
    @abstractmethod
    def deposit(self, bank_account_id: int, amount: float) -> bool:
        pass
    
    @abstractmethod
    def withdraw(self, bank_account_id: int, amount: float) -> bool:
        pass
    
    